feat#7 backend for email-generation

YashReddy7t opened the pull request:
This PR implements the backend logic for the Email Generator feature. The module allows AI-powered generation of email content based on candidate data and review verdicts.
Addresses #7

 What & Why
What:
Added a new backend module email_generator containing:

router.py: API endpoint /email-generator/ for generating email content.

schemas.py: Pydantic models for validating candidate data and request structure.

service.py: Business logic for formatting prompts and calling Gemini.

prompts.py: Gemini prompt template for generating emails.

__init__.py: Module init file.

Why:
To automate and standardize the email communication sent to candidates (e.g., rejection, selection, etc.) using structured prompts and LLMs (Google Gemini).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced an email generation API for HR recruiters to create professional candidate screening result emails.
  * Added support for structured candidate information, including name, role, experience, skills, verdict, rejection reasons, and HR notes.
  * Implemented a user-facing endpoint to generate tailored acceptance or rejection emails based on screening outcomes.

* **Bug Fixes**
  * Improved branch name validation script to accept "enhancement" as a valid branch type.

* **Style**
  * Minor formatting improvements for better code readability (no user-facing impact).

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SaiYashwanthReddyGujjula committed:
feat#7 backend for email-generation

sanjeevp-7t commented on backend/email_generator/router.py:
```diff
@@ -0,0 +1,19 @@
+router = APIRouter()
+
+
+@router.post("/generate")
+async def generate_email_endpoint(payload: EmailRequest):  # noqa: ANN201, D103
```
You don't need the `response_model` paramter, as FastAPI infers from the return type automatically.

Additionally, we are following the convention to the end the schemas with "Request" and "Response" suffixes. Check `schemas.py` in the job description folder.

Also, you should generally not ignore linter rules, unless you have a very good reason to.

```suggestion
+from backend.email_generator.schemas import GeneratedEmailRequest, GeneratedEmailResponse
+
+@router.post("/generate")
+async def generate_email(request: GeneratedEmailRequest) -> GeneratedEmailresponse:
```

SaiYashwanthReddyGujjula committed:
feat: integrate email generation functionality into backend.
Adresses #7
- Added email generator router and included it in the main API.
- Enhanced exception handling with custom LLMQuotaExceeded exception.
- Updated email request and response schemas for better validation and clarity.
- Implemented skills validation to ensure at least one skill is provided for candidates.
- Created an empty __init__.py file to make the email_generator directory a package.

SaiYashwanthReddyGujjula committed:
feat: integrate email generation functionality into backend.
Adresses #7
- Added email generator router and included it in the main API.
- Enhanced exception handling with custom LLMQuotaExceeded exception.
- Updated email request and response schemas for better validation and clarity.
- Implemented skills validation to ensure at least one skill is provided for candidates.
- Created an empty __init__.py file to make the email_generator directory a package.

SaiYashwanthReddyGujjula committed:
feat: integrate email generation functionality into backend.
Adresses #7
- Added email generator router and included it in the main API.
- Enhanced exception handling with custom LLMQuotaExceeded exception.
- Updated email request and response schemas for better validation and clarity.
- Implemented skills validation to ensure at least one skill is provided for candidates.
- Created an empty __init__.py file to make the email_generator directory a package.

SaiYashwanthReddyGujjula committed:
feat: integrate email generation functionality into backend.
Adresses #7
- Added email generator router and included it in the main API.
- Enhanced exception handling with custom LLMQuotaExceeded exception.
- Updated email request and response schemas for better validation and clarity.
- Implemented skills validation to ensure at least one skill is provided for candidates.
- Created an empty __init__.py file to make the email_generator directory a package.

YashReddy7t committed:
Merge branch 'develop' into feat/backend-7-email-generator-clean

Commit:
feat: add 'enhancement' type to branch name validation

SaiYashwanthReddyGujjula committed:
modified

SaiYashwanthReddyGujjula committed:
new

Alvitasy committed:
refactor(backend): migrate to structured output from Gemini

Replaced regex and manual JSON parsing with Gemini's generate_structured_output().
Introduced JobDescriptionLLMOutput schema for handling both job_description and error.
Removed extract_json_string() dependency and related parsing logic.
deleted utils.py
Refs: #23

SaiYashwanthReddyGujjula committed:
latest

SaiYashwanthReddyGujjula committed:
latest

SaiYashwanthReddyGujjula committed:
latest001

SaiYashwanthReddyGujjula committed:
latest002

Commit:
refactor(backend): simplify job description generation

- Removed LLMResponse model and integrated LLMErrorResponse for error handling.
- Updated job description generation to directly return the response from generate_structured_output.
- Cleaned up schemas and service logic for better clarity and maintainability.

Commit:
fix(gemini): update response parsing condition for clarity

- Changed the condition to check for `response.parsed` being `None` instead of using `not` for improved readability.
- Updated job description generation to return `JobDescriptionResponse` for better type consistency.

Commit:
fix(backend): rename JobDescriptionResult to JobDescriptionLLMResponse

sanjeevp-7t commented on temp_backup/router.py:
I'm not sure if this file should be committed.

sanjeevp-7t commented on backend/main.py:
```diff
@@ -2,19 +2,23 @@
 
-from backend.exceptions import APIException
+from backend.email_generator.router import router as email_generator_router
+from backend.exceptions import APIException  # from develop branch
```
Remove comment.
```suggestion
from backend.exceptions import APIException
```

sanjeevp-7t commented on backend/core/exceptions.py:
These newly added exceptions are redundant. Please remove them.

sanjeevp-7t commented on backend/core/gemini.py:
This file doesn't need any changes. Please undo.

sanjeevp-7t commented on backend/email_generator/__init__.py:
Remove comment. This file should be empty.

sanjeevp-7t commented on backend/email_generator/prompts.py:
```diff
@@ -0,0 +1,25 @@
+EMAIL_GENERATION_PROMPT = """
+You are an HR recruiter at Athena.7t. Generate a professional email to communicate the candidate's screening result.
```
Gemini won't know what is `Athena.7t`. Remove it.

```suggestion
You are an HR recruiter at a small-size tech company. Generate a professional email to communicate the candidate's screening result.
```

sanjeevp-7t commented on backend/email_generator/router.py:
```diff
@@ -0,0 +1,23 @@
+)
+async def generate_email_endpoint(request: EmailRequest) -> EmailGenerationResponse:
+    """Generate an email response based on the given candidate and verdict."""
+    with contextlib.suppress(Exception):
+        email_text = await generate_email(request)
+        return EmailGenerationResponse(generated_email=email_text)
```
I agree `contextlib` shouldn't be used here. However, there is a better convention to writing this file. The new `main.py` file handles all exception handling. The only thing neccessary for custom exceptions is to raise them and the FastAPI exception handler for `APIException` or `Exception` class will be handled accordingly.

There are also some inconsistencies with the current codebase conventions. Assume your service function returns an `EmailGenerationResponse`. Makes it cleaner.


```suggestion
from fastapi import APIRouter

from backend.email_generator.schemas import EmailGenerationResponse, EmailGenerationRequest
from backend.email_generator.service import generate_email

router = APIRouter(
    prefix="/email-generator",
    tags=["Email Generator"],
)


@router.post("/generate")
async def generate_candidate_email(request: EmailGenerationRequest) -> EmailGenerationResponse:
    """Generate an email response based on the given candidate and verdict."""
    return await generate_email(request)
```

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+        return v
+
+
+class EmailRequest(BaseModel):
```
```suggestion
class EmailGenerationRequest(BaseModel):
```

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+from pydantic import BaseModel, EmailStr, Field, field_validator
+
+
+class SkillsValidationError(ValueError):
+    """Exception raised when no skills are provided for a candidate."""
+
+    def __init__(self) -> None:
+        """Initialize the SkillsValidationError with a default message."""
+        super().__init__("At least one skill must be provided.")
```
This doesn't belong here. Remove this.

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+        super().__init__("At least one skill must be provided.")
+
+
+class CandidateInfo(BaseModel):
+    """Schema representing candidate information."""
+
+    name: str = Field(..., description="Full name of the candidate.")
+    email: EmailStr = Field(..., description="Email address of the candidate.")
+    skills: list[str] = Field(
+        ..., description="List of candidate's skills (at least one)."
+    )
+    experience: str = Field(..., description="Candidate's experience summary.")
+    title: str = Field(..., description="Job title the candidate applied for.")
+
+    @field_validator("skills")
+    def skills_must_have_at_least_one(cls, v: list[str]) -> list[str]:  # noqa: N805
+        """Ensure that the skills list contains at least one skill."""
+        if not v or len(v) < 1:
+            raise SkillsValidationError
+        return v
```
There is a built-in field validator for minimum length in Pydantic.


```suggestion
class CandidateInfo(BaseModel):
    """Schema representing candidate information."""

    name: str = Field(description="Full name of the candidate.")
    email: EmailStr = Field(description="Email address of the candidate.")
    skills: list[str] = Field(min_length=1, description="List of candidate's skills (at least one).")
    experience: str = Field(description="Candidate's experience summary.")
    title: str = Field(description="Job title the candidate applied for.")
```

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+    verdict: Literal["Yes", "No"] = Field(
+        ..., description='Final decision: "Yes" for selected, "No" for rejected.'
+    )
+    rejection_reason: str | None = Field(
+        None, description="Reason for rejection, if applicable."
+    )
```
Change this to `reason`. A user should be able to include a positive reason if a candidate passed, if they desire to.

```suggestion
    reason: str | None = Field(
        None, description="Reason for pass or rejection, if applicable."
    )
```

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+class EmailRequest(BaseModel):
+    """Schema representing an email generation request."""
+
+    candidate: CandidateInfo = Field(..., description="Details of the candidate.")
```
Don't need ellipsis.

```suggestion
    candidate: CandidateInfo = Field(description="Details of the candidate.")
```

sanjeevp-7t commented on backend/email_generator/schemas.py:
```diff
@@ -0,0 +1,51 @@
+class EmailGenerationResponse(BaseModel):
+    """Response model for a generated email."""
+
+    generated_email: str = Field(..., description="The generated email content.")
```
Don't need ellipsis.

```suggestion
    generated_email: str = Field(description="The generated email content.")
```

sanjeevp-7t commented on backend/email_generator/service.py:
```diff
@@ -0,0 +1,26 @@
+from backend.core.gemini import generate_text
+from backend.email_generator.prompts import EMAIL_GENERATION_PROMPT
+from backend.email_generator.schemas import EmailRequest
+
+
+async def generate_email(request: EmailRequest) -> str:
+    """Generate an email based on the EmailRequest object.
+
+    Args:
+        request (EmailRequest): Includes candidate info, verdict, rejection reason, and notes.
+
+    Returns:
+        str: The generated email text.
+
+    """
+    prompt = EMAIL_GENERATION_PROMPT.format(
+        name=request.candidate.name,
+        title=request.candidate.title,
+        experience=request.candidate.experience,
+        skills=", ".join(request.candidate.skills),
+        verdict=request.verdict,
+        rejection_reason=request.rejection_reason or "N/A",
+        notes=request.notes or "N/A",
+    )
+
+    return await generate_text(prompt)
```
Use Gemini's structured output, as it reduces code needed to get to your final response type.

```suggestion
from backend.core.gemini import generate_structured_output
from backend.email_generator.prompts import EMAIL_GENERATION_PROMPT
from backend.email_generator.schemas import EmailGenerationRequest, EmailGenerationResponse


async def generate_email(request: EmailGenerationRequest) -> EmailGenerationResponse:
    """Generate an candidate pass/rejection email."""
    prompt = EMAIL_GENERATION_PROMPT.format(
        name=request.candidate.name,
        title=request.candidate.title,
        experience=request.candidate.experience,
        skills=", ".join(request.candidate.skills),
        verdict=request.verdict,
        rejection_reason=request.rejection_reason
        notes=request.notes
    )
    return await generate_structured_output(prompt, EmailGenerationResponse)
```

sp7t reviewed (commented):
Also, remove `llm.py` from backend.core module

SaiYashwanthReddyGujjula committed:
latest012

SaiYashwanthReddyGujjula committed:
latest012

Commit:
docs: enhance CONTRIBUTING.md with GUI alternatives for Git operations

Commit:
docs: add enhancement issue template

SaiYashwanthReddyGujjula committed:
removed duplicate line in main
